### PR TITLE
refactor(frontend): synchronously acquire read snapshot

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Refactor the frontend snapshot manager to use `Arc` instead of manually tracking the query ID or transaction ID, so that pinning a snapshot can be simply done by `Arc::clone` without entering asynchronous context.

In this way, acquiring a read snapshot for a transaction can be called at any time, so that the epoch value obtained for `InlineNowProcTime` in the optimizer can be consistent with the snapshot for table scans. Therefore, the use of `NOW()` in both batch and streaming contexts (like Temporal Filter) will produce identical results in the same transaction.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [ ] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: e2e_test/batch/transaction/now.slt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+query T
+select count(*) > 0 from mv;
+----
+t
+
+# the result from batch query and materialized view should be the same
+query T
+select * from v
+except
+select * from mv;
```
Comment (BugenZhao) [inline e2e_test/batch/transaction/now.slt:33]: This test fails as we always yield the timestamp of the **next** epoch in `StreamNow` executor, before yielding the barrier message.

https://github.com/risingwavelabs/risingwave/blob/212cad4866ced33e1b1c960bfc1d8362e48d8239/src/stream/src/executor/now.rs#L98
https://github.com/risingwavelabs/risingwave/blob/212cad4866ced33e1b1c960bfc1d8362e48d8239/src/stream/src/executor/now.rs#L147

**Review by zwang28 — APPROVED** [review]
Rest LGTM

File: src/frontend/src/scheduler/snapshot.rs [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+    fn handle_operation(&mut self, operation: Operation) {
+        if operation.is_invalid() {
+            return;
+        }
+
+        match operation {
+            Operation::Pin(snapshot) => {
+                self.states
+                    .try_insert(SnapshotKey(snapshot), PinState::Pinned)
+                    .unwrap();
```
[the remark this answers is not in the document]
Comment (BugenZhao) [reply src/frontend/src/scheduler/snapshot.rs:320]: This seems like the key point that the current implementation works. If we have two identical `SnapshotKey` here, then we also have two corresponding `PinnedSnapshot` instances. In this way, we're not sure when both of them reach a reference count of zero, then unclear when to unpin.

Can we assert that `update` is always called with a strictly newer snapshot?

[the remark this answers is not in the document]
Comment (zwang28) [reply src/frontend/src/scheduler/snapshot.rs:320]: I think skipping new snapshot which LE than current latest_snapshot will do the job.

[the remark this answers is not in the document]
Comment (BugenZhao) [reply src/frontend/src/scheduler/snapshot.rs:321]: I've managed to make `update` more tolerable to ignore duplicate snapshots, but it still needs to assert the monotonicity according to the correctness mentioned above.

The `update` should only be called in a single-thread manner by the observer, and the monotonicity should already be checked on the meta service. So I guess this is not a big deal. 🤔

**Pull request merged**